{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index 0, ID = gi|142022655|gb|EQ086233.1|91, length 4635, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|91\n",
      "Name: gi|142022655|gb|EQ086233.1|91\n",
      "Description: gi|142022655|gb|EQ086233.1|91 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('CTCGCGTTGCAGGCCGGCGTGTCGCGCAACGACGTGTGGGGCCTGACGGGCAGG...GCC', SingleLetterAlphabet())\n",
      "Index 1, ID = gi|142022655|gb|EQ086233.1|304, length 1151, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|304\n",
      "Name: gi|142022655|gb|EQ086233.1|304\n",
      "Description: gi|142022655|gb|EQ086233.1|304 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('CGCGATCCAGTAGCGCTTGTAGCCGAGCGCTTCGGCACGCTTCGCGAGCGCGAT...GCC', SingleLetterAlphabet())\n",
      "Index 2, ID = gi|142022655|gb|EQ086233.1|255, length 4894, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|255\n",
      "Name: gi|142022655|gb|EQ086233.1|255\n",
      "Description: gi|142022655|gb|EQ086233.1|255 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('CTCGACGCGCTCCGCGTCGAGGTCGCCCGACGTCTCGCGCAGCAACTGATTCAA...TCG', SingleLetterAlphabet())\n",
      "Index 3, ID = gi|142022655|gb|EQ086233.1|45, length 3511, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|45\n",
      "Name: gi|142022655|gb|EQ086233.1|45\n",
      "Description: gi|142022655|gb|EQ086233.1|45 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('CGTGCTCGGCACGACTATCAGCCCGTATCTGTTTTTCTGGCAGGCCTCCCAGGA...CCG', SingleLetterAlphabet())\n",
      "Index 4, ID = gi|142022655|gb|EQ086233.1|396, length 4076, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|396\n",
      "Name: gi|142022655|gb|EQ086233.1|396\n",
      "Description: gi|142022655|gb|EQ086233.1|396 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('TGCATCGGCGACGGCTACTGGTTCCTCGAGAAAGCGCGCAACTACGCGAGCGAG...TCA', SingleLetterAlphabet())\n",
      "Index 5, ID = gi|142022655|gb|EQ086233.1|250, length 2867, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|250\n",
      "Name: gi|142022655|gb|EQ086233.1|250\n",
      "Description: gi|142022655|gb|EQ086233.1|250 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('GCCCGTGCCGACGCGCCGCATCGTCAGCGCGCGGCGCTCGTGGTCCTGCCACGC...CGA', SingleLetterAlphabet())\n",
      "Index 6, ID = gi|142022655|gb|EQ086233.1|322, length 442, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|322\n",
      "Name: gi|142022655|gb|EQ086233.1|322\n",
      "Description: gi|142022655|gb|EQ086233.1|322 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('TGCGGGCGGCCTGTGGGCGGTCGGCGCCGGGCTGTCGCAGCCGCTGTTCCACGG...GCG', SingleLetterAlphabet())\n",
      "Index 7, ID = gi|142022655|gb|EQ086233.1|88, length 890, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|88\n",
      "Name: gi|142022655|gb|EQ086233.1|88\n",
      "Description: gi|142022655|gb|EQ086233.1|88 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('GGACCGGCGAGGAAGTGGTCGGCGTCGAACAGGACGATCGCTGCGTCACGTTGC...CTC', SingleLetterAlphabet())\n",
      "Index 8, ID = gi|142022655|gb|EQ086233.1|594, length 967, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|594\n",
      "Name: gi|142022655|gb|EQ086233.1|594\n",
      "Description: gi|142022655|gb|EQ086233.1|594 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('GCCGCGGGCGCCGGACGGAGCACGCGCATGATCCCGCTTTCCCATCCGTCCTTC...TCG', SingleLetterAlphabet())\n",
      "Index 9, ID = gi|142022655|gb|EQ086233.1|293, length 4338, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|293\n",
      "Name: gi|142022655|gb|EQ086233.1|293\n",
      "Description: gi|142022655|gb|EQ086233.1|293 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('ACTGCCGCGTTCAGCGCGAGAATGTTCGTCTGGAACGCAATCCCTTCGATTACC...GAA', SingleLetterAlphabet())\n",
      "Index 10, ID = gi|142022655|gb|EQ086233.1|75, length 1352, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|75\n",
      "Name: gi|142022655|gb|EQ086233.1|75\n",
      "Description: gi|142022655|gb|EQ086233.1|75 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('TATAGGCCTGGACCTGCGTGCGGGCGAGATCGGCGTCTTCGCCGAAGGTGCCGT...GAG', SingleLetterAlphabet())\n",
      "Index 11, ID = gi|142022655|gb|EQ086233.1|454, length 4564, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|454\n",
      "Name: gi|142022655|gb|EQ086233.1|454\n",
      "Description: gi|142022655|gb|EQ086233.1|454 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('TCGCGGGCCTCGCGCTGCCGTTCGCGGCGCTCGCGTTGCTGGAGGTGGTCGTGC...CAG', SingleLetterAlphabet())\n",
      "Index 12, ID = gi|142022655|gb|EQ086233.1|16, length 4804, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|16\n",
      "Name: gi|142022655|gb|EQ086233.1|16\n",
      "Description: gi|142022655|gb|EQ086233.1|16 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('GTCGATCGACACGACGCTCGCGCAGCGCGACGCGAAGGCCGCGTGAGCGCACGA...CGT', SingleLetterAlphabet())\n",
      "Index 13, ID = gi|142022655|gb|EQ086233.1|584, length 964, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|584\n",
      "Name: gi|142022655|gb|EQ086233.1|584\n",
      "Description: gi|142022655|gb|EQ086233.1|584 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('CGCAGGTGGCAGCATTTAAGATCTCGCAAGGTGGTGTCTCCCCCATCATTGGTG...GGC', SingleLetterAlphabet())\n",
      "Index 14, ID = gi|142022655|gb|EQ086233.1|4, length 2095, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|4\n",
      "Name: gi|142022655|gb|EQ086233.1|4\n",
      "Description: gi|142022655|gb|EQ086233.1|4 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('ATGTGCCTGAAGACGTTTCAGGCGCGCGAATATTGTGGCGAGAGAAAAAATCTG...CAG', SingleLetterAlphabet())\n",
      "Index 15, ID = gi|142022655|gb|EQ086233.1|277, length 1432, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|277\n",
      "Name: gi|142022655|gb|EQ086233.1|277\n",
      "Description: gi|142022655|gb|EQ086233.1|277 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('CCGCGAAGGCCGCGTTCGCCACGCCCGCTGCCAACAGCGATCTCGCCGGCACCA...CGA', SingleLetterAlphabet())\n",
      "Index 16, ID = gi|142022655|gb|EQ086233.1|346, length 115, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|346\n",
      "Name: gi|142022655|gb|EQ086233.1|346\n",
      "Description: gi|142022655|gb|EQ086233.1|346 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('GCGGTCCCGGCGCCGCAGGCCGTCCGGCTCCTGCAGCGCGCCGAACCGGGTCTC...GCC', SingleLetterAlphabet())\n",
      "Index 17, ID = gi|142022655|gb|EQ086233.1|527, length 2646, with 0 features\n",
      "ID: gi|142022655|gb|EQ086233.1|527\n",
      "Name: gi|142022655|gb|EQ086233.1|527\n",
      "Description: gi|142022655|gb|EQ086233.1|527 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "Number of features: 0\n",
      "Seq('GAGAACCGGGAACCGGAACCATGACAGCCCCGCGCCGGTTTTACGCGAGATAGC...AGA', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "def get_info(fasta_file):\n",
    "    '''\n",
    "    Input: multi-fasta file\n",
    "    Output: \n",
    "    '''\n",
    "    for index, record in enumerate(SeqIO.parse(fasta_file, 'fasta')):\n",
    "        print('Index %i, ID = %s, length %i, with %i features' % \n",
    "              (index, record.id, len(record.seq), len(record.features)))\n",
    "        print(record)\n",
    "\n",
    "get_info('dna2.fasta')\n",
    "ex_dict = {'a':'ATGCGGTAGGTTTAA', 'b':'CGATGCGGATGCGCGCTATGCCCTAA'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n"
     ]
    }
   ],
   "source": [
    "def dict_seqs(fasta_file):\n",
    "    '''\n",
    "    Input: Fasta file \n",
    "    Output: Dictionary: key = sequence id, value = sequence from seqrecord\n",
    "    '''\n",
    "    seq_dict = {}\n",
    "    for index, record in enumerate(SeqIO.parse(fasta_file, 'fasta')):\n",
    "        seq_dict[record.id] = str(record.seq)\n",
    "    return seq_dict\n",
    "\n",
    "dictionary = dict_seqs('dna2.fasta')\n",
    "print(len(dictionary))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_seqlens(seq_dict):\n",
    "    '''\n",
    "    Input: Dictionary, key = sequence id, value = sequence from seqrecord\n",
    "    Output: Returns sorted list of tuples of seq id and sequence lengths from dictionary\n",
    "    '''    \n",
    "    seqlen_tuple = sorted((len(v),k) for (k,v) in seq_dict.items()) #list comprehension of tuple (k = id,v = length)\n",
    "    return seqlen_tuple\n",
    "\n",
    "s = sort_seqlens(dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Shortest: ', 115, 'Longest: ', 4894)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def len_minmax(seqs_tuple):\n",
    "    '''\n",
    "    Input: sorted list of tuples\n",
    "    Output: Returns max and min lengths of list of tuples\n",
    "    '''\n",
    "    longest = seqs_tuple[-1][0]\n",
    "    shortest = seqs_tuple[0][0]\n",
    "    return ('Shortest: ', shortest, 'Longest: ', longest)\n",
    "\n",
    "len_minmax(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(115, 'gi|142022655|gb|EQ086233.1|346')\n",
      "None count 1\n",
      "(442, 'gi|142022655|gb|EQ086233.1|322')\n",
      "None count 2\n",
      "(890, 'gi|142022655|gb|EQ086233.1|88')\n",
      "None count 3\n",
      "(964, 'gi|142022655|gb|EQ086233.1|584')\n",
      "None count 4\n",
      "(967, 'gi|142022655|gb|EQ086233.1|594')\n",
      "None count 5\n",
      "(1151, 'gi|142022655|gb|EQ086233.1|304')\n",
      "None count 6\n",
      "(1352, 'gi|142022655|gb|EQ086233.1|75')\n",
      "None count 7\n",
      "(1432, 'gi|142022655|gb|EQ086233.1|277')\n",
      "None count 8\n",
      "(2095, 'gi|142022655|gb|EQ086233.1|4')\n",
      "None count 9\n",
      "(2646, 'gi|142022655|gb|EQ086233.1|527')\n",
      "None count 10\n",
      "(2867, 'gi|142022655|gb|EQ086233.1|250')\n",
      "None count 11\n",
      "(3511, 'gi|142022655|gb|EQ086233.1|45')\n",
      "None count 12\n",
      "(4076, 'gi|142022655|gb|EQ086233.1|396')\n",
      "None count 13\n",
      "(4338, 'gi|142022655|gb|EQ086233.1|293')\n",
      "None count 14\n",
      "(4564, 'gi|142022655|gb|EQ086233.1|454')\n",
      "None count 15\n",
      "(4635, 'gi|142022655|gb|EQ086233.1|91')\n",
      "None count 16\n",
      "(4804, 'gi|142022655|gb|EQ086233.1|16')\n",
      "None count 17\n",
      "(4894, 'gi|142022655|gb|EQ086233.1|255')\n",
      "None count 18\n"
     ]
    }
   ],
   "source": [
    "def seq_length_dups(seq_tuples):\n",
    "    '''\n",
    "    Input: sorted list of tuples\n",
    "    Output: Prints sequences with the same lengths from a list of tuples with id and sequence lengths\n",
    "    '''\n",
    "    #print(seq_tuples)\n",
    "    len_dup = []\n",
    "    count = 0  \n",
    "    for i in seq_tuples:  #checks for duplication of lengths i[0]\n",
    "        print(i)\n",
    "        count += 1\n",
    "        if i[0] not in len_dup:\n",
    "            len_dup.append(i[0])\n",
    "            print('None', 'count',count)\n",
    "        \n",
    "        else:\n",
    "            print('Same length, Id:', i[1], 'length:', i[0], 'count',count)\n",
    "\n",
    "seq_length_dups(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rev(seq):\n",
    "    '''\n",
    "    Input: Sequence string\n",
    "    Output: Reverses sequence string\n",
    "    '''\n",
    "    seq_rev = seq[::-1]\n",
    "    return seq_rev\n",
    "\n",
    "def complement(seq):  \n",
    "    '''\n",
    "    Input: Sequence string\n",
    "    Output: Uses dictionary for complementary bases\n",
    "    '''\n",
    "    seq = seq.upper()\n",
    "    base_comp = {'A':'T','T':'A','C':'G','G':'C','N':'N'}\n",
    "    complement_seq = ''\n",
    "    for n in seq:\n",
    "        for key,value in base_comp.items():\n",
    "            if n == key:\n",
    "                complement_seq += value\n",
    "    return complement_seq\n",
    "\n",
    "def rev_complement(seq):\n",
    "    '''\n",
    "    Input: sequence string\n",
    "    Output: reverse complement of the input sequence\n",
    "    '''\n",
    "    seq = rev(seq)\n",
    "    seq = complement(seq)\n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a [['ATG', 'CGG', 'TAG', 'GTT', 'TAA'], ['TTA', 'AAC', 'CTA', 'CCG', 'CAT'], ['TGC', 'GGT', 'AGG', 'TTT', 'AA'], ['TAA', 'ACC', 'TAC', 'CGC', 'AT'], ['GCG', 'GTA', 'GGT', 'TTA', 'A'], ['AAA', 'CCT', 'ACC', 'GCA', 'T']]\n",
      "\n",
      "b [['CGA', 'TGC', 'GGA', 'TGC', 'GCG', 'CTA', 'TGC', 'CCT', 'AA'], ['TTA', 'GGG', 'CAT', 'AGC', 'GCG', 'CAT', 'CCG', 'CAT', 'CG'], ['GAT', 'GCG', 'GAT', 'GCG', 'CGC', 'TAT', 'GCC', 'CTA', 'A'], ['TAG', 'GGC', 'ATA', 'GCG', 'CGC', 'ATC', 'CGC', 'ATC', 'G'], ['ATG', 'CGG', 'ATG', 'CGC', 'GCT', 'ATG', 'CCC', 'TAA'], ['AGG', 'GCA', 'TAG', 'CGC', 'GCA', 'TCC', 'GCA', 'TCG']]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def frame_dict(seq_dict):\n",
    "    '''\n",
    "    Input: dictionary of sequences\n",
    "    Output: Returns a dictionary of all 6 frames (1_F,1_R,2_F,2_R,3_F,3_R) of each sequence in a list\n",
    "    '''\n",
    "    frames_dict = {}\n",
    "    \n",
    "    for key,seq in seq_dict.items():\n",
    "        #print(key)\n",
    "        revcomp_seq = rev_complement(seq_dict[key])  #reverse complement sequence\n",
    "        #print(revcomp_seq)\n",
    "        frames = []\n",
    "        \n",
    "        for j in range(0, 3):  #for 0,1,2 indeces: corresponds to frames\n",
    "            frame_f = []\n",
    "            frame_r = []\n",
    "            for n in range(j,len(seq_dict[key]),3):  #create codon for each frame\n",
    "                codon = seq_dict[key][n:n+3]\n",
    "                frame_f.append(codon)  #add codon to forward frame\n",
    "            \n",
    "            frames.append(frame_f)  #append each forward frame to list of frames\n",
    "            \n",
    "            for n in range(j,len(revcomp_seq),3):  #do the same but with rev comp\n",
    "                codon = revcomp_seq[n:n+3]\n",
    "                #print(codon)\n",
    "                frame_r.append(codon)\n",
    "            \n",
    "            frames.append(frame_r)\n",
    "        \n",
    "        frames_dict[key] = frames\n",
    "        continue\n",
    "        \n",
    "    return frames_dict\n",
    "\n",
    "fd = frame_dict(dictionary)\n",
    "#print(fd)\n",
    "ex_fd = frame_dict(ex_dict)\n",
    "for k,v in ex_fd.items():\n",
    "    print(k,v)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a\n",
      "A\n",
      "T\n",
      "G\n",
      "C\n",
      "0.267\n",
      "0.333\n",
      "0.333\n",
      "0.067\n",
      "[0.267, 0.333, 0.333, 0.067]\n",
      "Highest freq 1-mer is 0.333\n",
      "[0.067, 0.267, 0.333, 0.333]\n",
      "T\n",
      "A\n",
      "C\n",
      "G\n",
      "0.267\n",
      "0.333\n",
      "0.333\n",
      "0.067\n",
      "[0.267, 0.333, 0.333, 0.067]\n",
      "Highest freq 1-mer is 0.333\n",
      "[0.067, 0.267, 0.333, 0.333]\n",
      "T\n",
      "G\n",
      "C\n",
      "A\n",
      "0.357\n",
      "0.357\n",
      "0.071\n",
      "0.214\n",
      "[0.357, 0.357, 0.071, 0.214]\n",
      "Highest freq 1-mer is 0.357\n",
      "[0.071, 0.214, 0.357, 0.357]\n",
      "T\n",
      "A\n",
      "C\n",
      "G\n",
      "0.214\n",
      "0.357\n",
      "0.357\n",
      "0.071\n",
      "[0.214, 0.357, 0.357, 0.071]\n",
      "Highest freq 1-mer is 0.357\n",
      "[0.071, 0.214, 0.357, 0.357]\n",
      "G\n",
      "C\n",
      "T\n",
      "A\n",
      "0.385\n",
      "0.077\n",
      "0.308\n",
      "0.231\n",
      "[0.385, 0.077, 0.308, 0.231]\n",
      "Highest freq 1-mer is 0.385\n",
      "[0.077, 0.231, 0.308, 0.385]\n",
      "A\n",
      "C\n",
      "T\n",
      "G\n",
      "0.385\n",
      "0.385\n",
      "0.154\n",
      "0.077\n",
      "[0.385, 0.385, 0.154, 0.077]\n",
      "Highest freq 1-mer is 0.385\n",
      "[0.077, 0.154, 0.385, 0.385]\n",
      "b\n",
      "C\n",
      "G\n",
      "A\n",
      "T\n",
      "0.308\n",
      "0.308\n",
      "0.192\n",
      "0.192\n",
      "[0.308, 0.308, 0.192, 0.192]\n",
      "Highest freq 1-mer is 0.308\n",
      "[0.192, 0.192, 0.308, 0.308]\n",
      "T\n",
      "A\n",
      "G\n",
      "C\n",
      "0.192\n",
      "0.192\n",
      "0.308\n",
      "0.308\n",
      "[0.192, 0.192, 0.308, 0.308]\n",
      "Highest freq 1-mer is 0.308\n",
      "[0.192, 0.192, 0.308, 0.308]\n",
      "G\n",
      "A\n",
      "T\n",
      "C\n",
      "0.32\n",
      "0.2\n",
      "0.2\n",
      "0.28\n",
      "[0.32, 0.2, 0.2, 0.28]\n",
      "Highest freq 1-mer is 0.32\n",
      "[0.2, 0.2, 0.28, 0.32]\n",
      "T\n",
      "A\n",
      "G\n",
      "C\n",
      "0.16\n",
      "0.2\n",
      "0.32\n",
      "0.32\n",
      "[0.16, 0.2, 0.32, 0.32]\n",
      "Highest freq 1-mer is 0.32\n",
      "[0.16, 0.2, 0.32, 0.32]\n",
      "A\n",
      "T\n",
      "G\n",
      "C\n",
      "0.208\n",
      "0.208\n",
      "0.292\n",
      "0.292\n",
      "[0.208, 0.208, 0.292, 0.292]\n",
      "Highest freq 1-mer is 0.292\n",
      "[0.208, 0.208, 0.292, 0.292]\n",
      "A\n",
      "G\n",
      "C\n",
      "T\n",
      "0.208\n",
      "0.333\n",
      "0.333\n",
      "0.125\n",
      "[0.208, 0.333, 0.333, 0.125]\n",
      "Highest freq 1-mer is 0.333\n",
      "[0.125, 0.208, 0.333, 0.333]\n"
     ]
    }
   ],
   "source": [
    "def find_repeats(seq,kmer = 1): #default repeat length = 1\n",
    "    \n",
    "    '''\n",
    "    Input: string sequence; kmer = k-mer length\n",
    "    Output: dictionary key = repeat sequence of a certain kmer, value = repeat count\n",
    "    '''\n",
    "#iterate thru sequence, create 1) substrings list, 2) repeat count dictionary, and 3) a sequence list parsed by n for identifying repeats\n",
    "    \n",
    "    #print(seq)\n",
    "    substrings = []\n",
    "    subseq_count = {}  \n",
    "    seq_list_by_n =[]\n",
    "    \n",
    "    #create subsequences and add each unique one to substring list\n",
    "    for i in range(0,len(seq),kmer):\n",
    "        subseq = seq[i:i+kmer]\n",
    "        #print(subseq)\n",
    "        \n",
    "        if len(subseq) == kmer:\n",
    "            seq_list_by_n.append(subseq)  #list for counting subseqs\n",
    "            \n",
    "            if subseq not in substrings:  #list for identifying unique subseqs\n",
    "                substrings.append(subseq)\n",
    "            \n",
    "            else:\n",
    "                pass\n",
    "\n",
    "            #print(substrings)\n",
    "            \n",
    "            for s in substrings:  #add subseq to subseq_count dict as key\n",
    "                subseq_count[s] = 0  # start count at 0\n",
    "\n",
    "            for s in seq_list_by_n:  #count each repeat of subseq\n",
    "                subseq_count[s] += 1\n",
    "            \n",
    "        else: \n",
    "            pass\n",
    "\n",
    "    return subseq_count\n",
    "\n",
    "\n",
    "for key,seq in ex_fd.items():  #have to iterate thru dictionary \n",
    "        print(key)\n",
    "        repeat_list = []\n",
    "        for i in seq:\n",
    "            #print(i)\n",
    "            sequence = []\n",
    "            sequence += i\n",
    "            sequence = ''.join(sequence)\n",
    "            #print(sequence)\n",
    "            kmer = 1\n",
    "            repeats = find_repeats(sequence,kmer)  #remember 10 -> kmer length = 9\n",
    "            freq = repeat_freq(repeats)\n",
    "            print(freq)\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def repeat_freq(repeat_dict):  #default length of k-mer repeat = 1\n",
    "    '''\n",
    "    Input: repeat dictionary\n",
    "    Output: list of tuples (repeat sequence, frequency)\n",
    "    '''\n",
    "    repeat_list = []\n",
    "    total = 0\n",
    "    freq = []  #list of frequencies of each repeat\n",
    "    kmer = []   \n",
    "    \n",
    "    for k,v in repeat_dict.items():  #have to iterate thru dictionary\n",
    "        print(k)\n",
    "        kmer.append(len(k))\n",
    "        #print(kmer)\n",
    "        total += v\n",
    "    \n",
    "    for k,v in repeat_dict.items():\n",
    "        f = round((v/total),3)\n",
    "        print(f)\n",
    "        freq.append(f)\n",
    "    \n",
    "    print(freq)\n",
    "    s_freq = sorted(freq)\n",
    "    #print(s_freq)\n",
    "    print('Highest freq ' + str(kmer[1]) +'-mer ' + 'is ' + str(s_freq[-1]))\n",
    "    return s_freq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_start(l):\n",
    "    '''\n",
    "    Input: takes in a list of framed codons\n",
    "    Output: Returns index if a codon is id'd as start position in a list and returns a list of start indeces\n",
    "    '''\n",
    "    start = 'ATG'\n",
    "    start_list = []\n",
    "    for index,i in enumerate(l):  #order matters, index comes first then value of each item in list\n",
    "        if i == start:  #if codon is start codon\n",
    "            start_pos = index  #get starting index \n",
    "            start_list.append(start_pos)\n",
    "        else:\n",
    "            continue\n",
    "    return start_list\n",
    "    \n",
    "def find_stop(l):\n",
    "    '''\n",
    "    Input: a list of framed codons\n",
    "    Output: Function finds each stop position and returns a list of stop indeces\n",
    "    '''\n",
    "    stop = ['TAG', 'TGA', 'TAA']\n",
    "    stop_list = []\n",
    "    for index,i in enumerate(l):\n",
    "        #print(i)\n",
    "        for codon in stop:\n",
    "            if i == codon:  #for each stop codon in list\n",
    "                #print('y')\n",
    "                stop_pos = index + 1  #get stop index (+1 to include the stop codon)\n",
    "                #print('stop found', i)\n",
    "                stop_list.append(stop_pos)\n",
    "            else:\n",
    "                continue\n",
    "    #print(stop_list)\n",
    "    return stop_list\n",
    "\n",
    "for k,v in fd.items():\n",
    "    print(k)\n",
    "    for i in v:\n",
    "        print(i)\n",
    "        codon_list = i\n",
    "        find_stop(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def orfs(framed_dict):\n",
    "    '''\n",
    "    Input: Dictionary: key = frame, value = list of codons corresponding to frame\n",
    "    Output: Returns a nested dictionary, key0 = sequence id, value0 = dictionary, key1 = frame, value1 = sorted ([start,stop], orf, orf length)\n",
    "    '''\n",
    "    seqs = {}  #create dictionary to hold orfs of all frames per sequence\n",
    "    for key, value in framed_dict.items():\n",
    "        #print(key)\n",
    "        seqs[key] = []\n",
    "        orf_dict_list = []\n",
    "        #print(key,value)\n",
    "        frames = value  #assigns all frames for a sequence to variable frames\n",
    "        orf_dict = {}\n",
    "        for index,frame in enumerate(frames):  #for each of the 6 frames, create frame label\n",
    "            #print(orf_dict)\n",
    "            start = find_start(frame)  #find start indeces\n",
    "            stop = find_stop(frame)  #find stop indeces\n",
    "            \n",
    "            if index == 0:\n",
    "                f = '1_F'\n",
    "            if index == 1:\n",
    "                f = '1_R'\n",
    "            if index == 2:\n",
    "                f = '2_F'\n",
    "            if index == 3:\n",
    "                f = '2_R'\n",
    "            if index == 4:\n",
    "                f = '3_F'\n",
    "            if index == 5:\n",
    "                f = '3_R'\n",
    "            \n",
    "            orf_dict[f] = []\n",
    "            #print(f,'start',start,'stop',stop)\n",
    "            orfs = []\n",
    "            orf_l = []\n",
    "            indeces = []  #index list\n",
    "            #print(indeces)\n",
    "            orf_info = []\n",
    "            \n",
    "            \n",
    "            for i in start:\n",
    "                for j in stop:\n",
    "                    #print(i,j)\n",
    "                    if j > i:\n",
    "                        index = [i,j]\n",
    "                        #print(index)\n",
    "                        indeces.append(index)  #append indeces to list\n",
    "                        seq = frame[i:j]\n",
    "                        #print(f)\n",
    "                        orf = ''.join(seq)\n",
    "                        orf_l.append(len(orf))  #append orf length to orf_l list\n",
    "                        orfs.append(orf)  #append orf to orfs list\n",
    "                        #print(indeces,orfs,orf_l)        \n",
    "                    else:\n",
    "                        pass\n",
    "            \n",
    "            zipped = list(zip(indeces,orfs,orf_l))\n",
    "            #print(zipped)\n",
    "            orf_dict[f].append(sorted(zipped, key = lambda tup : tup[2]))   \n",
    "            #print(orf_dict)    \n",
    "            \n",
    "            seqs[key] = orf_dict\n",
    "        \n",
    "    return seqs  \n",
    "    \n",
    "\n",
    "print(ex_dict)\n",
    "orfs(fd)  #### problem with b, need the orfs to be in one list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## no longer needed. ^^ orfs returns sorted list \n",
    "def sorted_orfs(nested_dict):\n",
    "    '''\n",
    "    Input: nested dictionary, key = sequence id, value = (start,stop, orf, orf length) by frame \n",
    "    Output: nested dictionary, key = id, value = list of tuples (id,frame,start,stop,orf length) sorted by orf length\n",
    "    '''\n",
    "    s_orf_len = {}  #sorted orf lengths\n",
    "    for key,value in nested_dict.items():\n",
    "        #print(key,value)\n",
    "\n",
    "        for f,o in value.items():  # f = frame, o = start, stop, orf, orf_len\n",
    "            s = o[0][0]  #start\n",
    "            st = o[0][1]  #stop\n",
    "            orf = o[0][2]  #orf\n",
    "            l = o[0][3]\n",
    "            \n",
    "            if l:\n",
    "                print(o)\n",
    "                #print('l:',l)\n",
    "                l_zip = []  #list for each length\n",
    "                o_zip = []  #list for each orf\n",
    "                \n",
    "                for i in l: #for each length in l\n",
    "                    #print('i:',i)\n",
    "                    l_zip.append(i)\n",
    "                    for j in orf:  #for each orf \n",
    "                        if len(j) == i:  #if length of orf corresponds to length item \n",
    "                            #print('j:',j, 'i:',i)\n",
    "                            o_zip.append(j)\n",
    "                        else:\n",
    "                            pass\n",
    "                        \n",
    "                #print(l_zip,o_zip)  \n",
    "                z = list(zip(o_zip,l_zip))\n",
    "                print(z)\n",
    "                \n",
    "                z_f = list(zip(s,st,z))\n",
    "                print(z_f)\n",
    "            \n",
    "          \n",
    "            #z_sorted = [a,b,c,d for a,b,c,d in sorted(zip(s,st,o_zip,l_zip))]\n",
    "            \n",
    "            \n",
    "            \n",
    "        #s_orf_len[key] = (f,zipped)\n",
    "    #return s_orf_len\n",
    "\n",
    "\n",
    "sorted_orfs(b)\n",
    "#print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dict_to_list(sorteddict_):\n",
    "    '''\n",
    "    Input: dictionary, key = sequence id, value = sorted list of tuples ('t') (id, frame, orf length)\n",
    "    Output: list of sorted lists of tuples\n",
    "    '''\n",
    "    orflist = []\n",
    "    for k,v in sorteddict_:\n",
    "        l.append(v)\n",
    "    return orflist\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''def orf_len_by_frame(sorted_orf_dict):\n",
    "    \n",
    "    Input: dictionary, key = sequence id, value = sorted list of tuples ('t') (id,frame, orf length), user input: frame to subset list by\n",
    "    Output: Sorted list of orf lengths in requested frame\n",
    "    \n",
    "    l = sorted_orf_dict\n",
    "    frame_list = []\n",
    "    f_r = input('What frame do you want? (1_F, 1_R, 2_F, 2_R, 3_F, 3_R)')\n",
    "    for k,v in l.items():\n",
    "        #print(k,v)\n",
    "        for t in v:\n",
    "            if t[0] == f_r:\n",
    "                frame_list.append(t[1])\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "    return frame_list'''\n",
    "\n",
    "def orf_len_by_frame(orflist):\n",
    "    '''\n",
    "    Input: list of tuples (id,frame,orf length)\n",
    "    Output: a list of tuples subset by frame\n",
    "    '''\n",
    "    subset_l = []\n",
    "    frame = input('What frame do you want? (1_F, 1_R, 2_F, 2_R, 3_F, 3_R)')\n",
    "    for i in orflist:\n",
    "        if orflist[1] == frame:\n",
    "            subset_l.append(i)\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "    return subset_l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''def orf_len_by_id(sorted_orf_dict):\n",
    "    \n",
    "    Input:\n",
    "    Output:\n",
    "    \n",
    "    l = sorted_orf_dict\n",
    "    frame_list = []\n",
    "    f_r = input('What frame do you want? (1_F, 1_R, 2_F, 2_R, 3_F, 3_R)')\n",
    "    for k,v in l.items():\n",
    "        #print(k,v)\n",
    "        for t in v:\n",
    "            if t[0] == f_r:\n",
    "                frame_list.append(t[1])\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "    return frame_list'''\n",
    "\n",
    "def orf_len_by_id(orflist):\n",
    "    '''\n",
    "    Input: list of list of tuples (id,frame,orf length)\n",
    "    Output: a list of tuples subset by frame\n",
    "    '''\n",
    "    subset_l = []\n",
    "    id_ = input('What id do you want?')\n",
    "    for i in orflist:\n",
    "        if orflist[0] == id_:\n",
    "            subset_l.append(i)\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "    return subset_l\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summary(fasta):\n",
    "    dictionary = dict_seqs(fasta)\n",
    "    lengths = sort_seqlens(dictionary)  #get sequence lengths and sort them into a list\n",
    "    print('Number of sequences:', len(lengths)) \n",
    "    print('Length of sequences', lengths) \n",
    "    print('Smallest and longest sequences:', len_minmax(lengths)) \n",
    "    print('Are there multiple sequences with the same lengths?', seq_dups(lengths))\n",
    "    print('Are there repeats in the sequences?', repeats(dictionary))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#starting position of the longest ORF in reading frame 3 in any of the sequences? ATG = position 1\n",
    "\n",
    "b = orfs(fd)\n",
    "#print(b)\n",
    "\n",
    "def longest_orf_framedir(nested_orf_dict, frame = 1, direction = 'F'):\n",
    "    \n",
    "    '''\n",
    "    Input: nested dictionary, key = seq id, value = sorted orfs and info by frame; frame = ['1','2','3'] (default = 1); direction = 'F','R' or 'B' (default = 'F')\n",
    "    Output: longest orf with its info (seq id,[start,stop],orf sequence, length)\n",
    "    '''\n",
    "    \n",
    "    longest_orfs = []\n",
    "    #print(frame,direction)\n",
    "    for i in frame:\n",
    "        #print(i)\n",
    "        for k,v in b.items():\n",
    "            #print(k,v)\n",
    "            for k1,v1 in v.items():\n",
    "                #print(k1,v1)\n",
    "                if k1[0] == i and k1[2] == direction:\n",
    "                    #print(i,direction)\n",
    "                    #print(k1,v1)\n",
    "                    for x in v1:\n",
    "                        if x:\n",
    "                            #print(x)\n",
    "                            longest_orfs.append((k,x[-1]))\n",
    "                elif k1[0] == i and direction == 'B':\n",
    "                    #print(k1,v1)\n",
    "                    for y in v1:\n",
    "                        if y:\n",
    "                            #print(y)\n",
    "                            longest_orfs.append((k,y[-1]))\n",
    "\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "\n",
    "    #print(longest_orfs)\n",
    "    if longest_orfs:\n",
    "        longest_orf = sorted(longest_orfs, key = lambda tup : tup[1][2])[-1]\n",
    "        return longest_orf\n",
    "    \n",
    "    else:\n",
    "        print(\"No orfs in this frame and direction!\")\n",
    "\n",
    "longest_orf(b, frame = ['1'], direction = 'F')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## What is the length of the longest ORF appearing in any sequence and in any forward reading frame?\n",
    "\n",
    "if \n",
    "longest_orf(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## What is the length of the longest forward ORF that appears in the sequence with the identifier gi|142022655|gb|EQ086233.1|16?\n",
    "\n",
    "_id = input('Id?')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Find the most frequently occurring repeat of length 6 in all sequences. How many times does it occur in all?\n",
    "\n",
    "print(repeat_freq(##framed_dict##, 6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Find all repeats of length 12 in the input file. Let's use Max to specify the number of copies of the most frequent repeat of length 12. How many different 12-base sequences occur Max times?\n",
    "\n",
    "mfr = repeat_freq(#framed_dict, 12)\n",
    "Max = mfr[-1][1]\n",
    "\n",
    "for i in mfr:\n",
    "    if i[1] == max:\n",
    "        print(i[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Which one of the following repeats of length 7 has a maximum number of occurrences?\n",
    "\n",
    "a = repeat_freq(#framed_dict, 7)\n",
    "repeat = a[-1][0]\n",
    "print(repeat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ex_dict = {'a':'ATGCGGTAGGTTAA', 'b':'CGATGCGGATGCGCGCTTAA'}\n",
    "\n",
    "\n",
    "print(ex_dict)\n",
    "#summary(ex_dict)\n",
    "\n",
    "frames = frame_dict(ex_dict)  #create frames from dictionary\n",
    "orfs = orfs(frames)  #find orfs in each frame\n",
    "sorted_orfs = sorted_orfs(orfs)  #get lengths and id of orfs \n",
    "#orf_list = dict_to_list(sorted_orfs)\n",
    "#subset_frame = orf_len_by_frame(orf_list)\n",
    "#subset_id_frame = orf_len_by_id(subset_frame)\n",
    "\n",
    "#print(orf_len_frame[-1])  #print longest orf in this frame\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
